blog/content/post/2020-08-05-PCIe-features.md

```diff
@@ -0,0 +1,177 @@
+
+## Introduction
+
+Every device in the PCIe tree has at least one function. Each function has its
+configuration space. Configuration space is device registers mapped to either
+I/O memory space (used mostly by PCI devices) or to normal memory space. PCI
+capable configuration space has 256 bytes, extended configuration space used in
+PCIe has 4096 bytes.
+
+In this memory we can find most importantly 2 structures.
+
+- **Header**  
+![config header](/img/Pci-config-space.png)  
+It contains information like vendor ID, and device ID, pointers to memory
+assigned to the function and pointer to first
+capability in linked list.
+- **Linked list of capabilities**  
+Each capability has its own header, it contains ID of the
+capability, its version and pointer to the next capability in the list. Beneath
+this header are mapped registers needed by capabilities.
+
+To check if a given capability is implemented by function, a software has to
+search through the list and check if a given capability ID is present in it.
+
+## What PCIe capability actually is
+
+It is nothing more than a predefined feature of the device's function, a feature
+that is known to be possible to be implemented in all devices, but most of them
+are optional. Every capability has registers in a device mapped to configuration
+space. Those registers are a kind of interface to the capability. Some are read
+only and contain information needed by the software, others are read/write and
+can be used to pass some information to the function.
+
+## List of capabilities structures:
+(This list can be incomplete)
+
+- **PCI Express Capability register block**
+
+This is one of the most important capability structures, it must be present in
+all PCIe functions. It contains information about:
+
+1. Device e.g.:  
+Maximum supported payload size, maximum accepted latency
+2. Link e.g.:  
+Link max speed, link max width, link latency, port number for given link
+3. Port e.g.:  
+Physical slot number, Hot-Plug capability
+
+**Hot-Plug**
+
+Registers responsible for this capability are located in the Capability register
+block. This capability allows change of device in the PCIe slot at runtime.
+
+**Baseline Error Reporting**
+
+Bits corresponding to this capability are located partly in the configuration
+header, partly in the Capability register block. All of them must be present in
+all PCIe functions.
+
+- **Power Management**
+
+It permits the OS to manage the power environment of a Function directly by
+accessing registers corresponding to this capability. OS can set device stat to
+one of 4 states. 2 of them are mandatory, and 2 can be optionally implemented.
+D0 state is a state where the function is fully operational and uses full power,
+every next state (D1,D2,D3) uses lower power, but also takes more time to recover
+to D0.
+
+- **Message Signaled Interrupts**
+
+PCI has pins to let the central interrupt controller know that it needs service.
```

Comment:
MSI is a write to memory, it does not use wired signals or pins. This is true in case of legacy INTx